In [1]:
#Import NYC Airbnb Dataset
#Import Python Libraries
import pandas as pd
import numpy as np

#Read the Files
hotel_info = 'C:/Users/DELL/OneDrive/Desktop/Coursera/DATASETS/Airbnb Dataset/NYC Airbnb Dataset/NYC_Airbnb_Hotel_Info.csv'

hotel = pd.read_csv(hotel_info)

In [2]:
#host info 
host_info = 'C:/Users/DELL/OneDrive/Desktop/Coursera/DATASETS/Airbnb Dataset/NYC Airbnb Dataset/NYC_Airbnb_Host_Info.csv'

host = pd.read_csv(host_info)

In [3]:
#location_info
location_info = 'C:/Users/DELL/OneDrive/Desktop/Coursera/DATASETS/Airbnb Dataset/NYC Airbnb Dataset/NYC_Airbnb_Neighbourhood_Info.csv'

location = pd.read_csv(location_info)

In [4]:
#review info
review_info = 'C:/Users/DELL/OneDrive/Desktop/Coursera/DATASETS/Airbnb Dataset/NYC Airbnb Dataset/NYC_Airbnb_Reviews_Info.csv'

review = pd.read_csv(review_info)

C:\Users\DELL\anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3185: DtypeWarning: Columns (13) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [5]:
#Verify if data has been read successfully.
hotel.head()

id  host_id                        listing_url     scrape_id  \
0  2595.0     2845  https://www.airbnb.com/rooms/2595  2.022120e+13   
1  5203.0     7490  https://www.airbnb.com/rooms/5203  2.022120e+13   
2  5136.0     7378  https://www.airbnb.com/rooms/5136  2.022120e+13   
3  5121.0     7356  https://www.airbnb.com/rooms/5121  2.022120e+13   
4  6848.0    15991  https://www.airbnb.com/rooms/6848  2.022120e+13   

  last_scraped           source                                      name  \
0   05-12-2022      city scrape                     Skylit Midtown Castle   
1   05-12-2022  previous scrape        Cozy Clean Guest Room - Family Apt   
2   04-12-2022      city scrape  Spacious Brooklyn Duplex, Patio + Garden   
3   05-12-2022      city scrape                           BlissArtsSpace!   
4   05-12-2022      city scrape          Only 2 stops to Manhattan studio   

                                         description  \
0  Beautiful, spacious skylit studio in the heart...   
1  Our best guests are seeking a safe, clean, spa...   
2  We welcome you to stay in our lovely 2 br dupl...   
3  One room available for rent in a 2 bedroom apt...   
4  Comfortable studio apartment with super comfor...   

                               neighborhood_overview  \
0  Centrally located in the heart of Manhattan ju...   
1  Our neighborhood is full of restaurants and ca...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                                         picture_url            neighbourhood  \
0  https://a0.muscache.com/pictures/f0813a11-40b2...  New York, United States   
1  https://a0.muscache.com/pictures/103776/b37157...  New York, United States   
2  https://a0.muscache.com/pictures/miso/Hosting-...                      NaN   
3  https://a0.muscache.com/pictures/2090980c-b68e...                      NaN   
4  https://a0.muscache.com/pictures/e4f031a7-f146...                      NaN   

   number_of_reviews  
0                 49  
1                118  
2                  3  
3                 50  
4                187

In [6]:
host.head()

id  host_id                                 host_url      host_name  \
0  2595.0     2845   https://www.airbnb.com/users/show/2845       Jennifer   
1  5203.0     7490   https://www.airbnb.com/users/show/7490      MaryEllen   
2  5136.0     7378   https://www.airbnb.com/users/show/7378        Rebecca   
3  5121.0     7356   https://www.airbnb.com/users/show/7356          Garon   
4  6848.0    15991  https://www.airbnb.com/users/show/15991  Allen & Irina   

   host_since host_location  \
0  09-09-2008  New York, NY   
1  05-02-2009  New York, NY   
2  03-02-2009  New York, NY   
3  03-02-2009  New York, NY   
4  06-05-2009  New York, NY   

                                          host_about host_response_time  \
0  A New Yorker since (Phone number hidden by Air...       within a day   
1  Welcome to family life with my oldest two away...                NaN   
2  Rebecca is an artist/designer, and Henoch is i...                NaN   
3   I am an artist(painter, filmmaker) and curato...       within a day   
4  We love to travel. When we travel we like to s...     within an hour   

  host_response_rate host_acceptance_rate host_is_superhost  \
0                72%                  22%                 f   
1                NaN                  NaN                 f   
2                NaN                  50%                 f   
3                90%                  82%                 t   
4               100%                 100%                 f   

                                  host_thumbnail_url  \
0  https://a0.muscache.com/im/pictures/user/50fc5...   
1  https://a0.muscache.com/im/users/7490/profile_...   
2  https://a0.muscache.com/im/users/7378/profile_...   
3  https://a0.muscache.com/im/pictures/user/72a61...   
4  https://a0.muscache.com/im/users/15991/profile...   

                                    host_picture_url  host_neighbourhood  \
0  https://a0.muscache.com/im/pictures/user/50fc5...             Midtown   
1  https://a0.muscache.com/im/users/7490/profile_...     Upper West Side   
2  https://a0.muscache.com/im/users/7378/profile_...   Greenwood Heights   
3  https://a0.muscache.com/im/pictures/user/72a61...  Bedford-Stuyvesant   
4  https://a0.muscache.com/im/users/15991/profile...        Williamsburg   

   host_listings_count  host_total_listings_count  \
0                  6.0                        9.0   
1                  1.0                        5.0   
2                  1.0                        5.0   
3                  2.0                        2.0   
4                  1.0                        1.0   

                 host_verifications host_has_profile_pic  \
0  ['email', 'phone', 'work_email']                    t   
1                ['email', 'phone']                    t   
2                ['email', 'phone']                    t   
3                ['email', 'phone']                    t   
4                ['email', 'phone']                    t   

  host_identity_verified  
0                      t  
1                      t  
2                      t  
3                      t  
4                      t

In [7]:
location.head()

id            neighbourhood neighbourhood_cleansed  \
0  2595.0  New York, United States                Midtown   
1  5203.0  New York, United States        Upper West Side   
2  5136.0                      NaN            Sunset Park   
3  5121.0                      NaN     Bedford-Stuyvesant   
4  6848.0                      NaN           Williamsburg   

  neighbourhood_group_cleansed  latitude  longitude  \
0                    Manhattan  40.75356  -73.98559   
1                    Manhattan  40.80380  -73.96751   
2                     Brooklyn  40.66265  -73.99454   
3                     Brooklyn  40.68535  -73.95512   
4                     Brooklyn  40.70935  -73.95342   

                 property_type        room_type  accommodates  bathrooms  \
0           Entire rental unit  Entire home/apt             1        NaN   
1  Private room in rental unit     Private room             1        NaN   
2           Entire rental unit  Entire home/apt             4        NaN   
3  Private room in rental unit     Private room             2        NaN   
4           Entire rental unit  Entire home/apt             3        NaN   

  bathrooms_text  bedrooms  beds  \
0         1 bath       NaN   1.0   
1  1 shared bath       1.0   1.0   
2      1.5 baths       2.0   2.0   
3            NaN       1.0   1.0   
4         1 bath       NaN   1.0   

                                           amenities    price  
0  ["Carbon monoxide alarm", "Stove", "Bathtub", ...  $175.00  
1  ["Hair dryer", "Smoke alarm", "Carbon monoxide...   $75.00  
2  ["HDTV with Disney+, HBO Max, standard cable, ...  $275.00  
3  ["Heating", "Wifi", "Long term stays allowed",...   $60.00  
4  ["Carbon monoxide alarm", "Shampoo", "Stove", ...   $68.00

In [8]:
review.head()

id  number_of_reviews  number_of_reviews_ltm  number_of_reviews_l30d  \
0  2595.0                 49                      1                       0   
1  5203.0                118                      0                       0   
2  5136.0                  3                      1                       0   
3  5121.0                 50                      0                       0   
4  6848.0                187                      5                       0   

  first_review last_review  review_scores_rating  review_scores_accuracy  \
0   21-11-2009  21-06-2022                  4.68                    4.73   
1   07-09-2009  21-07-2017                  4.91                    4.83   
2   02-01-2014  10-08-2022                  5.00                    5.00   
3   28-05-2009  02-12-2019                  4.52                    4.22   
4   25-05-2009  02-11-2022                  4.57                    4.59   

   review_scores_cleanliness  review_scores_checkin  \
0                       4.63                   4.77   
1                       4.82                   4.97   
2                       5.00                   5.00   
3                       4.09                   4.91   
4                       4.86                   4.84   

   review_scores_communication  review_scores_location  review_scores_value  \
0                         4.80                    4.81                 4.40   
1                         4.95                    4.94                 4.92   
2                         5.00                    4.67                 5.00   
3                         4.91                    4.47                 4.52   
4                         4.80                    4.67                 4.56   

  license instant_bookable  calculated_host_listings_count  \
0     NaN                f                               3   
1     NaN                f                               1   
2     NaN                f                               1   
3     NaN                f                               2   
4     NaN                f                               1   

   calculated_host_listings_count_entire_homes  \
0                                            3   
1                                            0   
2                                            1   
3                                            0   
4                                            1   

   calculated_host_listings_count_private_rooms  \
0                                             0   
1                                             1   
2                                             0   
3                                             2   
4                                             0   

   calculated_host_listings_count_shared_rooms  reviews_per_month  
0                                            0               0.31  
1                                            0               0.73  
2                                            0               0.03  
3                                            0               0.30  
4                                            0               1.13

In [9]:
# Remove uneccessary columns from each table.
#hotel_info
hotel.drop("listing_url", axis=1, inplace=True)
hotel.drop("neighborhood_overview", axis=1, inplace=True)
hotel.drop("picture_url", axis=1, inplace=True)
hotel.head()

id  host_id     scrape_id last_scraped           source  \
0  2595.0     2845  2.022120e+13   05-12-2022      city scrape   
1  5203.0     7490  2.022120e+13   05-12-2022  previous scrape   
2  5136.0     7378  2.022120e+13   04-12-2022      city scrape   
3  5121.0     7356  2.022120e+13   05-12-2022      city scrape   
4  6848.0    15991  2.022120e+13   05-12-2022      city scrape   

                                       name  \
0                     Skylit Midtown Castle   
1        Cozy Clean Guest Room - Family Apt   
2  Spacious Brooklyn Duplex, Patio + Garden   
3                           BlissArtsSpace!   
4          Only 2 stops to Manhattan studio   

                                         description            neighbourhood  \
0  Beautiful, spacious skylit studio in the heart...  New York, United States   
1  Our best guests are seeking a safe, clean, spa...  New York, United States   
2  We welcome you to stay in our lovely 2 br dupl...                      NaN   
3  One room available for rent in a 2 bedroom apt...                      NaN   
4  Comfortable studio apartment with super comfor...                      NaN   

   number_of_reviews  
0                 49  
1                118  
2                  3  
3                 50  
4                187

In [10]:
# host_info
host.drop("host_url", axis=1, inplace=True)
host.drop("host_about", axis=1, inplace=True)
host.drop("host_thumbnail_url", axis=1, inplace=True)
host.drop("host_picture_url", axis=1, inplace=True)
host.drop("host_neighbourhood", axis=1, inplace=True)
host.head()

id  host_id      host_name  host_since host_location  \
0  2595.0     2845       Jennifer  09-09-2008  New York, NY   
1  5203.0     7490      MaryEllen  05-02-2009  New York, NY   
2  5136.0     7378        Rebecca  03-02-2009  New York, NY   
3  5121.0     7356          Garon  03-02-2009  New York, NY   
4  6848.0    15991  Allen & Irina  06-05-2009  New York, NY   

  host_response_time host_response_rate host_acceptance_rate  \
0       within a day                72%                  22%   
1                NaN                NaN                  NaN   
2                NaN                NaN                  50%   
3       within a day                90%                  82%   
4     within an hour               100%                 100%   

  host_is_superhost  host_listings_count  host_total_listings_count  \
0                 f                  6.0                        9.0   
1                 f                  1.0                        5.0   
2                 f                  1.0                        5.0   
3                 t                  2.0                        2.0   
4                 f                  1.0                        1.0   

                 host_verifications host_has_profile_pic  \
0  ['email', 'phone', 'work_email']                    t   
1                ['email', 'phone']                    t   
2                ['email', 'phone']                    t   
3                ['email', 'phone']                    t   
4                ['email', 'phone']                    t   

  host_identity_verified  
0                      t  
1                      t  
2                      t  
3                      t  
4                      t

In [11]:
#location_info
location.drop("neighbourhood_cleansed", axis=1, inplace=True)
location.head()

id            neighbourhood neighbourhood_group_cleansed  latitude  \
0  2595.0  New York, United States                    Manhattan  40.75356   
1  5203.0  New York, United States                    Manhattan  40.80380   
2  5136.0                      NaN                     Brooklyn  40.66265   
3  5121.0                      NaN                     Brooklyn  40.68535   
4  6848.0                      NaN                     Brooklyn  40.70935   

   longitude                property_type        room_type  accommodates  \
0  -73.98559           Entire rental unit  Entire home/apt             1   
1  -73.96751  Private room in rental unit     Private room             1   
2  -73.99454           Entire rental unit  Entire home/apt             4   
3  -73.95512  Private room in rental unit     Private room             2   
4  -73.95342           Entire rental unit  Entire home/apt             3   

   bathrooms bathrooms_text  bedrooms  beds  \
0        NaN         1 bath       NaN   1.0   
1        NaN  1 shared bath       1.0   1.0   
2        NaN      1.5 baths       2.0   2.0   
3        NaN            NaN       1.0   1.0   
4        NaN         1 bath       NaN   1.0   

                                           amenities    price  
0  ["Carbon monoxide alarm", "Stove", "Bathtub", ...  $175.00  
1  ["Hair dryer", "Smoke alarm", "Carbon monoxide...   $75.00  
2  ["HDTV with Disney+, HBO Max, standard cable, ...  $275.00  
3  ["Heating", "Wifi", "Long term stays allowed",...   $60.00  
4  ["Carbon monoxide alarm", "Shampoo", "Stove", ...   $68.00

In [12]:
#reviews_info
review.drop("calculated_host_listings_count", axis=1, inplace=True)
review.drop("calculated_host_listings_count_entire_homes", axis=1, inplace=True)
review.drop("calculated_host_listings_count_private_rooms", axis=1, inplace=True)
review.drop("calculated_host_listings_count_shared_rooms", axis=1, inplace=True)
review.drop("review_scores_accuracy", axis=1, inplace=True)
review.head()

id  number_of_reviews  number_of_reviews_ltm  number_of_reviews_l30d  \
0  2595.0                 49                      1                       0   
1  5203.0                118                      0                       0   
2  5136.0                  3                      1                       0   
3  5121.0                 50                      0                       0   
4  6848.0                187                      5                       0   

  first_review last_review  review_scores_rating  review_scores_cleanliness  \
0   21-11-2009  21-06-2022                  4.68                       4.63   
1   07-09-2009  21-07-2017                  4.91                       4.82   
2   02-01-2014  10-08-2022                  5.00                       5.00   
3   28-05-2009  02-12-2019                  4.52                       4.09   
4   25-05-2009  02-11-2022                  4.57                       4.86   

   review_scores_checkin  review_scores_communication  review_scores_location  \
0                   4.77                         4.80                    4.81   
1                   4.97                         4.95                    4.94   
2                   5.00                         5.00                    4.67   
3                   4.91                         4.91                    4.47   
4                   4.84                         4.80                    4.67   

   review_scores_value license instant_bookable  reviews_per_month  
0                 4.40     NaN                f               0.31  
1                 4.92     NaN                f               0.73  
2                 5.00     NaN                f               0.03  
3                 4.52     NaN                f               0.30  
4                 4.56     NaN                f               1.13

In [13]:
# Find the Missing Data from each csv file inserted.
#host info
missing_host = host.isnull()
missing_host.head()

for column in missing_host.columns.values.tolist():
    print(column)
    print(missing_host[column].value_counts())
    print(" ")

id
False    41533
Name: id, dtype: int64
 
host_id
False    41533
Name: host_id, dtype: int64
 
host_name
False    41528
True         5
Name: host_name, dtype: int64
 
host_since
False    41528
True         5
Name: host_since, dtype: int64
 
host_location
False    33788
True      7745
Name: host_location, dtype: int64
 
host_response_time
False    27888
True     13645
Name: host_response_time, dtype: int64
 
host_response_rate
False    27888
True     13645
Name: host_response_rate, dtype: int64
 
host_acceptance_rate
False    29322
True     12211
Name: host_acceptance_rate, dtype: int64
 
host_is_superhost
False    41504
True        29
Name: host_is_superhost, dtype: int64
 
host_listings_count
False    41528
True         5
Name: host_listings_count, dtype: int64
 
host_total_listings_count
False    41528
True         5
Name: host_total_listings_count, dtype: int64
 
host_verifications
False    41533
Name: host_verifications, dtype: int64
 
host_has_profile_pic
False    41528
True     

# Host_Info

host_is_superhost -> 29 missing values

host_listings_count -> 5 missing values

host_total_listings_count -> 5 missing values

host_has_profile_pic -> 5 missing values

host_identity_verified -> 5 missing values

host_name -> 5 missing values


host_since -> 5 missing values

In [14]:
#Replace Missing Values with Values 
#host_is_superhost
max_superhost = host["host_is_superhost"].value_counts().idxmax()
host["host_is_superhost"].replace(np.NaN, max_superhost, inplace=True)
host.head()

#host_listings_count
avg_host_list_count = host["host_listings_count"].mean(axis=0)
host["host_listings_count"].replace(np.NaN, avg_host_list_count, inplace=True)
host.head()

#host_total_listings_count
avg_host_totlist_count = host["host_total_listings_count"].mean(axis=0)
host["host_total_listings_count"].replace(np.NaN, avg_host_totlist_count, inplace=True)
host.head()

#host_has_profile_pic
max_hostprofile = host["host_has_profile_pic"].value_counts().idxmax()
host["host_has_profile_pic"].replace(np.NaN, max_hostprofile, inplace=True)
host.head()

#host_has_profile_pic
max_hostidentity = host["host_identity_verified"].value_counts().idxmax()
host["host_identity_verified"].replace(np.NaN, max_hostidentity, inplace=True)
host.head()


id  host_id      host_name  host_since host_location  \
0  2595.0     2845       Jennifer  09-09-2008  New York, NY   
1  5203.0     7490      MaryEllen  05-02-2009  New York, NY   
2  5136.0     7378        Rebecca  03-02-2009  New York, NY   
3  5121.0     7356          Garon  03-02-2009  New York, NY   
4  6848.0    15991  Allen & Irina  06-05-2009  New York, NY   

  host_response_time host_response_rate host_acceptance_rate  \
0       within a day                72%                  22%   
1                NaN                NaN                  NaN   
2                NaN                NaN                  50%   
3       within a day                90%                  82%   
4     within an hour               100%                 100%   

  host_is_superhost  host_listings_count  host_total_listings_count  \
0                 f                  6.0                        9.0   
1                 f                  1.0                        5.0   
2                 f                  1.0                        5.0   
3                 t                  2.0                        2.0   
4                 f                  1.0                        1.0   

                 host_verifications host_has_profile_pic  \
0  ['email', 'phone', 'work_email']                    t   
1                ['email', 'phone']                    t   
2                ['email', 'phone']                    t   
3                ['email', 'phone']                    t   
4                ['email', 'phone']                    t   

  host_identity_verified  
0                      t  
1                      t  
2                      t  
3                      t  
4                      t

In [15]:
#host_response_rate
host["host_response_rate"] = host["host_response_rate"].str.rstrip("%").astype("float")/100.00
avg_host_resp_rate = host["host_response_rate"].mean(axis=0)
host["host_response_rate"].replace(np.NaN, avg_host_resp_rate, inplace=True)
host.head()

#host_acceptance_rate
host["host_acceptance_rate"] = host["host_acceptance_rate"].str.rstrip("%").astype("float")/100.00
avg_host_accp_rate = host["host_acceptance_rate"].mean(axis=0)
host["host_acceptance_rate"].replace(np.NaN, avg_host_accp_rate, inplace=True)
host.head()

#host_response_time
max_host_resp_time = host["host_response_time"].value_counts().idxmax()
host["host_response_time"].replace(np.NaN, max_host_resp_time, inplace=True)
host.head()

#host_location
max_host_loc = host["host_location"].value_counts().idxmax()
host["host_location"].replace(np.NaN, max_host_loc, inplace=True)
host.head()

id  host_id      host_name  host_since host_location  \
0  2595.0     2845       Jennifer  09-09-2008  New York, NY   
1  5203.0     7490      MaryEllen  05-02-2009  New York, NY   
2  5136.0     7378        Rebecca  03-02-2009  New York, NY   
3  5121.0     7356          Garon  03-02-2009  New York, NY   
4  6848.0    15991  Allen & Irina  06-05-2009  New York, NY   

  host_response_time  host_response_rate  host_acceptance_rate  \
0       within a day            0.720000              0.220000   
1     within an hour            0.943671              0.828656   
2     within an hour            0.943671              0.500000   
3       within a day            0.900000              0.820000   
4     within an hour            1.000000              1.000000   

  host_is_superhost  host_listings_count  host_total_listings_count  \
0                 f                  6.0                        9.0   
1                 f                  1.0                        5.0   
2                 f                  1.0                        5.0   
3                 t                  2.0                        2.0   
4                 f                  1.0                        1.0   

                 host_verifications host_has_profile_pic  \
0  ['email', 'phone', 'work_email']                    t   
1                ['email', 'phone']                    t   
2                ['email', 'phone']                    t   
3                ['email', 'phone']                    t   
4                ['email', 'phone']                    t   

  host_identity_verified  
0                      t  
1                      t  
2                      t  
3                      t  
4                      t

In [16]:
# Drop Missing Values
#host_name -> 5 missing values & #host_since -> 5 missing values
host.dropna(subset=["host_name"], axis=0, inplace=True)
host.reset_index(drop=True, inplace=True)

In [17]:
#location info
missing_location = location.isnull()
missing_location.head()

for column in missing_location.columns.values.tolist():
    print(column)
    print(missing_location[column].value_counts())
    print(" ")


id
False    41533
Name: id, dtype: int64
 
neighbourhood
False    24090
True     17443
Name: neighbourhood, dtype: int64
 
neighbourhood_group_cleansed
False    41533
Name: neighbourhood_group_cleansed, dtype: int64
 
latitude
False    41533
Name: latitude, dtype: int64
 
longitude
False    41533
Name: longitude, dtype: int64
 
property_type
False    41533
Name: property_type, dtype: int64
 
room_type
False    41533
Name: room_type, dtype: int64
 
accommodates
False    41533
Name: accommodates, dtype: int64
 
bathrooms
True    41533
Name: bathrooms, dtype: int64
 
bathrooms_text
False    41456
True        77
Name: bathrooms_text, dtype: int64
 
bedrooms
False    37711
True      3822
Name: bedrooms, dtype: int64
 
beds
False    40592
True       941
Name: beds, dtype: int64
 
amenities
False    41533
Name: amenities, dtype: int64
 
price
False    41533
Name: price, dtype: int64
 


# Location_info

neighbourhood -> neighbourhood has 17443 missing values

bathrooms -> bathrooms has 41533 missing values

bathrooms_text -> bathrooms_text has 77 missing values

bedrooms -> bedrooms has 3822 missing values

beds -> beds has 941 missing values


In [18]:
location.tail()

id neighbourhood neighbourhood_group_cleansed   latitude  \
41528  7.726830e+17           NaN                        Bronx  40.870936   
41529  7.727050e+17           NaN                       Queens  40.698912   
41530  7.727110e+17           NaN                    Manhattan  40.802590   
41531  7.727140e+17           NaN                        Bronx  40.828074   
41532  7.727170e+17           NaN                    Manhattan  40.740030   

       longitude                property_type        room_type  accommodates  \
41528 -73.900649           Entire rental unit  Entire home/apt             4   
41529 -73.804666         Private room in home     Private room             2   
41530 -73.962380  Private room in rental unit     Private room             2   
41531 -73.847886           Entire rental unit  Entire home/apt             4   
41532 -73.997302           Entire rental unit  Entire home/apt             6   

       bathrooms    bathrooms_text  bedrooms  beds  \
41528        NaN            1 bath       2.0   2.0   
41529        NaN    1 private bath       1.0   1.0   
41530        NaN  1.5 shared baths       1.0   1.0   
41531        NaN            1 bath       2.0   2.0   
41532        NaN           2 baths       2.0   3.0   

                                               amenities      price  
41528  ["Smoke alarm", "Wifi", "Exercise equipment", ...    $105.00  
41529  ["Smoke alarm", "First aid kit", "Wifi", "TV",...     $87.00  
41530  ["Hair dryer", "Smoke alarm", "Carbon monoxide...     $70.00  
41531  ["Smoke alarm", "Carbon monoxide alarm", "Firs...    $125.00  
41532  ["Carbon monoxide alarm", "Shampoo", "Conditio...  $1,114.00

In [19]:
# neighbourhood
location["neighbourhood"] = "New York, United States"
location.head()

# Replacing the Hotel field neighbourhood to neighbourhood_group_cleansed
#first to drop neighbourhood from hotel_info
hotel.drop("neighbourhood", axis=1, inplace=True)
hotel.head()

hotel = hotel.join(location["neighbourhood_group_cleansed"])
hotel.head()

location.head()

id            neighbourhood neighbourhood_group_cleansed  latitude  \
0  2595.0  New York, United States                    Manhattan  40.75356   
1  5203.0  New York, United States                    Manhattan  40.80380   
2  5136.0  New York, United States                     Brooklyn  40.66265   
3  5121.0  New York, United States                     Brooklyn  40.68535   
4  6848.0  New York, United States                     Brooklyn  40.70935   

   longitude                property_type        room_type  accommodates  \
0  -73.98559           Entire rental unit  Entire home/apt             1   
1  -73.96751  Private room in rental unit     Private room             1   
2  -73.99454           Entire rental unit  Entire home/apt             4   
3  -73.95512  Private room in rental unit     Private room             2   
4  -73.95342           Entire rental unit  Entire home/apt             3   

   bathrooms bathrooms_text  bedrooms  beds  \
0        NaN         1 bath       NaN   1.0   
1        NaN  1 shared bath       1.0   1.0   
2        NaN      1.5 baths       2.0   2.0   
3        NaN            NaN       1.0   1.0   
4        NaN         1 bath       NaN   1.0   

                                           amenities    price  
0  ["Carbon monoxide alarm", "Stove", "Bathtub", ...  $175.00  
1  ["Hair dryer", "Smoke alarm", "Carbon monoxide...   $75.00  
2  ["HDTV with Disney+, HBO Max, standard cable, ...  $275.00  
3  ["Heating", "Wifi", "Long term stays allowed",...   $60.00  
4  ["Carbon monoxide alarm", "Shampoo", "Stove", ...   $68.00

In [20]:
# Bathrooms
location["bathrooms"] = location["bathrooms_text"].str.extractall('([-+]?\d*\.?\d+)').unstack().astype("float")
#location["bathrooms"] = location["bathrooms_text"].explode()
location.head()

# bathrooms_text
max_bathroom = location["bathrooms_text"].value_counts().idxmax()
location["bathrooms_text"].replace(np.NaN, max_bathroom, inplace=True)
location.head()
# I am leaving some blanks as there is a chance that we can have no vacancies in a hotel.

id            neighbourhood neighbourhood_group_cleansed  latitude  \
0  2595.0  New York, United States                    Manhattan  40.75356   
1  5203.0  New York, United States                    Manhattan  40.80380   
2  5136.0  New York, United States                     Brooklyn  40.66265   
3  5121.0  New York, United States                     Brooklyn  40.68535   
4  6848.0  New York, United States                     Brooklyn  40.70935   

   longitude                property_type        room_type  accommodates  \
0  -73.98559           Entire rental unit  Entire home/apt             1   
1  -73.96751  Private room in rental unit     Private room             1   
2  -73.99454           Entire rental unit  Entire home/apt             4   
3  -73.95512  Private room in rental unit     Private room             2   
4  -73.95342           Entire rental unit  Entire home/apt             3   

   bathrooms bathrooms_text  bedrooms  beds  \
0        1.0         1 bath       NaN   1.0   
1        1.0  1 shared bath       1.0   1.0   
2        1.5      1.5 baths       2.0   2.0   
3        NaN         1 bath       1.0   1.0   
4        1.0         1 bath       NaN   1.0   

                                           amenities    price  
0  ["Carbon monoxide alarm", "Stove", "Bathtub", ...  $175.00  
1  ["Hair dryer", "Smoke alarm", "Carbon monoxide...   $75.00  
2  ["HDTV with Disney+, HBO Max, standard cable, ...  $275.00  
3  ["Heating", "Wifi", "Long term stays allowed",...   $60.00  
4  ["Carbon monoxide alarm", "Shampoo", "Stove", ...   $68.00

In [21]:
#Bedrooms
max_beds = location["beds"].value_counts().idxmax()
location["beds"].replace(np.NaN, max_beds, inplace=True)
location.head()

id            neighbourhood neighbourhood_group_cleansed  latitude  \
0  2595.0  New York, United States                    Manhattan  40.75356   
1  5203.0  New York, United States                    Manhattan  40.80380   
2  5136.0  New York, United States                     Brooklyn  40.66265   
3  5121.0  New York, United States                     Brooklyn  40.68535   
4  6848.0  New York, United States                     Brooklyn  40.70935   

   longitude                property_type        room_type  accommodates  \
0  -73.98559           Entire rental unit  Entire home/apt             1   
1  -73.96751  Private room in rental unit     Private room             1   
2  -73.99454           Entire rental unit  Entire home/apt             4   
3  -73.95512  Private room in rental unit     Private room             2   
4  -73.95342           Entire rental unit  Entire home/apt             3   

   bathrooms bathrooms_text  bedrooms  beds  \
0        1.0         1 bath       NaN   1.0   
1        1.0  1 shared bath       1.0   1.0   
2        1.5      1.5 baths       2.0   2.0   
3        NaN         1 bath       1.0   1.0   
4        1.0         1 bath       NaN   1.0   

                                           amenities    price  
0  ["Carbon monoxide alarm", "Stove", "Bathtub", ...  $175.00  
1  ["Hair dryer", "Smoke alarm", "Carbon monoxide...   $75.00  
2  ["HDTV with Disney+, HBO Max, standard cable, ...  $275.00  
3  ["Heating", "Wifi", "Long term stays allowed",...   $60.00  
4  ["Carbon monoxide alarm", "Shampoo", "Stove", ...   $68.00

In [22]:
#Review info
missing_review = review.isnull()
missing_review.head()

for column in missing_review.columns.values.tolist():
    print(column)
    print(missing_review[column].value_counts())
    print(" ")


id
False    41533
Name: id, dtype: int64
 
number_of_reviews
False    41533
Name: number_of_reviews, dtype: int64
 
number_of_reviews_ltm
False    41533
Name: number_of_reviews_ltm, dtype: int64
 
number_of_reviews_l30d
False    41533
Name: number_of_reviews_l30d, dtype: int64
 
first_review
False    32140
True      9393
Name: first_review, dtype: int64
 
last_review
False    32140
True      9393
Name: last_review, dtype: int64
 
review_scores_rating
False    32140
True      9393
Name: review_scores_rating, dtype: int64
 
review_scores_cleanliness
False    31702
True      9831
Name: review_scores_cleanliness, dtype: int64
 
review_scores_checkin
False    31688
True      9845
Name: review_scores_checkin, dtype: int64
 
review_scores_communication
False    31697
True      9836
Name: review_scores_communication, dtype: int64
 
review_scores_location
False    31685
True      9848
Name: review_scores_location, dtype: int64
 
review_scores_value
False    31685
True      9848
Name: review_sco

# Reviews Info

first_review -> first_review has 9393 missing values

last_review -> last_review has 9393 missing values

review_scores_rating -> review_scores_rating has 9393 missing values

review_scores_cleanliness -> review_scores_cleanliness has 9831 missing values

review_scores_checkin -> review_scores_checkin has 9845 missing values

review_scores_communication -> review_scores_communication has 9836 missing values

review_scores_location -> review_scores_location has 9848 missing values

review_scores_value -> review_scores_value has 9848 missing values

license -> license has 1 missing value

reviews_per_month -> reviews_per_month has 9393 missing values

In [23]:
#Reviews
review["review_scores_rating"].replace(np.NaN, 0, inplace=True)
review["reviews_per_month"].replace(np.NaN, 0, inplace=True)
review.head()

id  number_of_reviews  number_of_reviews_ltm  number_of_reviews_l30d  \
0  2595.0                 49                      1                       0   
1  5203.0                118                      0                       0   
2  5136.0                  3                      1                       0   
3  5121.0                 50                      0                       0   
4  6848.0                187                      5                       0   

  first_review last_review  review_scores_rating  review_scores_cleanliness  \
0   21-11-2009  21-06-2022                  4.68                       4.63   
1   07-09-2009  21-07-2017                  4.91                       4.82   
2   02-01-2014  10-08-2022                  5.00                       5.00   
3   28-05-2009  02-12-2019                  4.52                       4.09   
4   25-05-2009  02-11-2022                  4.57                       4.86   

   review_scores_checkin  review_scores_communication  review_scores_location  \
0                   4.77                         4.80                    4.81   
1                   4.97                         4.95                    4.94   
2                   5.00                         5.00                    4.67   
3                   4.91                         4.91                    4.47   
4                   4.84                         4.80                    4.67   

   review_scores_value license instant_bookable  reviews_per_month  
0                 4.40     NaN                f               0.31  
1                 4.92     NaN                f               0.73  
2                 5.00     NaN                f               0.03  
3                 4.52     NaN                f               0.30  
4                 4.56     NaN                f               1.13

In [24]:
#Fix the First_review and Last_review date
max_fdate = review["first_review"].value_counts().idxmax()
review["first_review"].replace(np.NaN,max_fdate, inplace=True)
review.head()

#Last_review date
max_ldate = review["last_review"].value_counts().idxmax()
review["last_review"].replace(np.NaN,max_ldate, inplace=True)
review.head()

id  number_of_reviews  number_of_reviews_ltm  number_of_reviews_l30d  \
0  2595.0                 49                      1                       0   
1  5203.0                118                      0                       0   
2  5136.0                  3                      1                       0   
3  5121.0                 50                      0                       0   
4  6848.0                187                      5                       0   

  first_review last_review  review_scores_rating  review_scores_cleanliness  \
0   21-11-2009  21-06-2022                  4.68                       4.63   
1   07-09-2009  21-07-2017                  4.91                       4.82   
2   02-01-2014  10-08-2022                  5.00                       5.00   
3   28-05-2009  02-12-2019                  4.52                       4.09   
4   25-05-2009  02-11-2022                  4.57                       4.86   

   review_scores_checkin  review_scores_communication  review_scores_location  \
0                   4.77                         4.80                    4.81   
1                   4.97                         4.95                    4.94   
2                   5.00                         5.00                    4.67   
3                   4.91                         4.91                    4.47   
4                   4.84                         4.80                    4.67   

   review_scores_value license instant_bookable  reviews_per_month  
0                 4.40     NaN                f               0.31  
1                 4.92     NaN                f               0.73  
2                 5.00     NaN                f               0.03  
3                 4.52     NaN                f               0.30  
4                 4.56     NaN                f               1.13

In [25]:
#review_scores_cleanliness 
review["review_scores_cleanliness"].replace(np.NaN, 0, inplace=True)
#review_scores_checkin 
review["review_scores_checkin"].replace(np.NaN, 0, inplace=True)
#review_scores_communication
review["review_scores_communication"].replace(np.NaN, 0, inplace=True)
#review_scores_location 
review["review_scores_location"].replace(np.NaN, 0, inplace=True)
#review_scores_value 
review["review_scores_value"].replace(np.NaN, 0, inplace=True)
review.tail()

id  number_of_reviews  number_of_reviews_ltm  \
41528  7.726830e+17                  0                      0   
41529  7.727050e+17                  0                      0   
41530  7.727110e+17                  0                      0   
41531  7.727140e+17                  0                      0   
41532  7.727170e+17                  0                      0   

       number_of_reviews_l30d first_review last_review  review_scores_rating  \
41528                       0   05-09-2022  20-11-2022                   0.0   
41529                       0   05-09-2022  20-11-2022                   0.0   
41530                       0   05-09-2022  20-11-2022                   0.0   
41531                       0   05-09-2022  20-11-2022                   0.0   
41532                       0   05-09-2022  20-11-2022                   0.0   

       review_scores_cleanliness  review_scores_checkin  \
41528                        0.0                    0.0   
41529                        0.0                    0.0   
41530                        0.0                    0.0   
41531                        0.0                    0.0   
41532                        0.0                    0.0   

       review_scores_communication  review_scores_location  \
41528                          0.0                     0.0   
41529                          0.0                     0.0   
41530                          0.0                     0.0   
41531                          0.0                     0.0   
41532                          0.0                     0.0   

       review_scores_value license instant_bookable  reviews_per_month  
41528                  0.0     NaN                f                0.0  
41529                  0.0     NaN                t                0.0  
41530                  0.0     NaN                t                0.0  
41531                  0.0     NaN                t                0.0  
41532                  0.0     NaN                t                0.0

In [26]:
#Verify their Data Types and fix the formats

host.dtypes
#change the host_id -> to be object
#chane host_since -> to be datetime
#change host_response_rate -> to be float64
#change host_acceptance_rate -> to be float64

location.dtypes
#change price -> to be float

review.dtypes
#change first_review -> to be datetime
#change last_review -> to be datetime

#hotel.dtypes
#change hotel id -> to be object
#change hotel last_scrapped -> to be datetime 

id                             float64
number_of_reviews                int64
number_of_reviews_ltm            int64
number_of_reviews_l30d           int64
first_review                    object
last_review                     object
review_scores_rating           float64
review_scores_cleanliness      float64
review_scores_checkin          float64
review_scores_communication    float64
review_scores_location         float64
review_scores_value            float64
license                         object
instant_bookable                object
reviews_per_month              float64
dtype: object

In [27]:
# Formatting the Data Types
hotel.head()

# Id - must be object
hotel[["id"]] = hotel[["id"]].astype("object")
# host_id - must be object
hotel[["host_id"]] = hotel[["host_id"]].astype("object")
# scrape_id - must be object
hotel[["scrape_id"]] = hotel[["scrape_id"]].astype("object")
# last_scraped - must be datetime
#hotel["last_scraped"] = pd.to_datetime(hotel["last_scraped"])
hotel.dtypes

id                              object
host_id                         object
scrape_id                       object
last_scraped                    object
source                          object
name                            object
description                     object
number_of_reviews                int64
neighbourhood_group_cleansed    object
dtype: object

In [28]:
host[["id","host_id"]] = host[["id","host_id"]].astype(object)
host.dtypes

id                            object
host_id                       object
host_name                     object
host_since                    object
host_location                 object
host_response_time            object
host_response_rate           float64
host_acceptance_rate         float64
host_is_superhost             object
host_listings_count          float64
host_total_listings_count    float64
host_verifications            object
host_has_profile_pic          object
host_identity_verified        object
dtype: object

In [29]:
#neighbourhood data types
location.dtypes

# id to be object
#location[["id"]] = location[["id"]].astype("object")
# accommodates in int
#location[["accommodates"]] = location[["accommadates"]].astype("int")
#price, bathrooms, bedrooms, beds in float
#location[["bathrooms","bedrooms","beds"]] = location[["bathrooms","bedrooms","beds"]].astype("float")

location["price"] = location["price"].replace('[\$,]', '', regex=True).astype("float")
location.dtypes

id                              float64
neighbourhood                    object
neighbourhood_group_cleansed     object
latitude                        float64
longitude                       float64
property_type                    object
room_type                        object
accommodates                      int64
bathrooms                       float64
bathrooms_text                   object
bedrooms                        float64
beds                            float64
amenities                        object
price                           float64
dtype: object

In [30]:
#reviews
review[["id"]] = review[["id"]].astype("object")

In [31]:
review.dtypes

id                              object
number_of_reviews                int64
number_of_reviews_ltm            int64
number_of_reviews_l30d           int64
first_review                    object
last_review                     object
review_scores_rating           float64
review_scores_cleanliness      float64
review_scores_checkin          float64
review_scores_communication    float64
review_scores_location         float64
review_scores_value            float64
license                         object
instant_bookable                object
reviews_per_month              float64
dtype: object

In [32]:
#remove duplicates values from file

#hotel info
hotel.duplicated().sum()

hotel.drop_duplicates(inplace=True)

hotel.duplicated().sum()

0

In [33]:
# host info
host.duplicated().sum()

host.drop_duplicates(inplace=True)

host.duplicated().sum()

0

In [34]:
# location info

location.duplicated().sum()

location.drop_duplicates(inplace=True)

location.duplicated().sum()

0

In [35]:
# review info
review.duplicated().sum()

review.drop_duplicates(inplace=True)

review.duplicated().sum()

0

In [36]:
len(hotel)

41527

In [37]:
len(host)

41204

In [38]:
len(location)

41530

In [39]:
len(review)

41206

In [40]:
hotel_file_path = 'C:/Users/DELL/OneDrive/Desktop/Coursera/DATASETS/Airbnb Dataset/NYC Airbnb Dataset/NYC_Airbnb_Hotel_Info_updated.xlsx'
hotel.to_excel(hotel_file_path)

In [41]:
host_file_path = 'C:/Users/DELL/OneDrive/Desktop/Coursera/DATASETS/Airbnb Dataset/NYC Airbnb Dataset/NYC_Airbnb_Host_Info_updated.xlsx'
host.to_excel(host_file_path)

In [42]:
location_file_path = 'C:/Users/DELL/OneDrive/Desktop/Coursera/DATASETS/Airbnb Dataset/NYC Airbnb Dataset/NYC_Airbnb_Neighbourhood_Info_updated.xlsx'
location.to_excel(location_file_path)

In [43]:
review_file_path = 'C:/Users/DELL/OneDrive/Desktop/Coursera/DATASETS/Airbnb Dataset/NYC Airbnb Dataset/NYC_Airbnb_Reviews_Info_updated.xlsx'
review.to_excel(review_file_path)